File: test/language/templates/cpp98/InstantiateTypeAsSqlTableFieldTest.cpp
#include "gtest/gtest.h"

#include "zserio/SqliteConnection.h"

#include "templates/instantiate_type_as_sql_table_field/Test32Table.h"

namespace templates
{
namespace instantiate_type_as_sql_table_field
{

static const char SQLITE3_MEM_DB[] = ":memory:";

TEST(InstantiateSqlTableTest, instantiationOfTest32Table)
{
    sqlite3 *internalConnection = NULL;
    int result = sqlite3_open(SQLITE3_MEM_DB, &internalConnection);
    ASSERT_EQ(SQLITE_OK, result);
    zserio::SqliteConnection connection(internalConnection, ::zserio::SqliteConnection::INTERNAL_CONNECTION);

    Test32Table test32Table(connection, "test32Table");
    test32Table.createTable();

    std::vector<Test32TableRow> rows;
    Test32TableRow row;
    row.setId(13);
    Test32 test32;
    test32.setValue(42);
    row.setTest(test32);
    rows.push_back(row);
    test32Table.write(rows);

    std::vector<Test32TableRow> readRows;
    test32Table.read(readRows);
    ASSERT_EQ(1, readRows.size());

    ASSERT_EQ(13, readRows.at(0).getId());
    ASSERT_EQ(42, readRows.at(0).getTest().getValue());
}

} // namespace instantiate_type_as_sql_table_field
} // namespace templates

File: test/language/templates/java/templates/InstantiateTypeAsSqlTableFieldTest.java
package templates;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Properties;
import java.util.List;
import java.util.ArrayList;

import zserio.runtime.BitSizeOfCalculator;
import zserio.runtime.ZserioError;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.BitStreamWriter;
import zserio.runtime.io.FileBitStreamReader;
import zserio.runtime.io.FileBitStreamWriter;

import test_utils.JdbcUtil;

import templates.instantiate_type_as_sql_table_field.Test32Table;
import templates.instantiate_type_as_sql_table_field.Test32TableRow;
import templates.instantiate_type_as_sql_table_field.Test32;

public class InstantiateTypeAsSqlTableFieldTest
{
    @BeforeAll
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Test
    public void readWrite() throws IOException, SQLException
    {
        final Properties connectionProps = new Properties();
        connectionProps.setProperty("flags", "CREATE");
        final String uriPath = "jdbc:sqlite:" + SQLITE_MEM_DB;

        final Connection connection = DriverManager.getConnection(uriPath, connectionProps);
        final Test32Table test32Table = new Test32Table(connection, "test32Table");
        test32Table.createTable();

        final Test32TableRow row = new Test32TableRow();
        row.setId(13);
        row.setTest(new Test32(42));
        final List<Test32TableRow> rows = new ArrayList<Test32TableRow>();
        rows.add(row);
        test32Table.write(rows);

        final List<Test32TableRow> readRows = test32Table.read();
        assertEquals(1, readRows.size());

        assertEquals(13, readRows.get(0).getId());
        assertEquals(42, readRows.get(0).getTest().getValue());
    }

    private static final String SQLITE_MEM_DB = ":memory:";
}
